### DISCUSSION TITLE
uiHelper.RendererCallback can not be triggered

### DISCUSSION THREAD

**Discussion by yicong123456** [google/filament#7144] [Q&A]
I tried to use filament in one of fragment in android project followed sample code and tried as these
 private val binding get() = _binding!!
 override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Filament.init();
        choreographer = Choreographer.getInstance()
        // frameScheduler = FrameCallback()
 }
override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
          Log.i("oncreateview", "80");

        _binding = FragmentFilamentBinding.inflate(inflater, container, false);

        choreographer = Choreographer.getInstance()
        displayHelper = DisplayHelper(requireActivity())

        setupSurfaceView()
        engine = Engine.create()
        renderer = engine!!.createRenderer();
        scene = engine!!.createScene();
        camera = engine!!.createCamera(engine!!.entityManager.create());
        view = engine!!.createView()
        setupView()
        setupScene()
        return binding.root
    }

in the setupSurfaceView:

private fun setupSurfaceView() {
        surfaceView = binding.glSurfaceView;
        uiHelper = UiHelper(UiHelper.ContextErrorPolicy.DONT_CHECK);

        //uiHelper.renderCallback = SurfaceCallback();
        uiHelper.renderCallback = object : UiHelper.RendererCallback {
            override fun onNativeWindowChanged(surface: Surface) {
                Log.i("surface in callback", surface.toString());
                swapChain?.let { engine?.destroySwapChain(it) }
                Log.i("surface in callback", surface.toString());
                swapChain = engine?.createSwapChain(surface)
                Log.i("swapchain is: ", swapChain.toString());
                displayHelper.attach(renderer, surfaceView.display)
            }

            override fun onDetachedFromSurface() {
                displayHelper.detach()
                swapChain?.let {
                    engine?.destroySwapChain(it)
                    // Required to ensure we don't return before Filament is done executing the
                    // destroySwapChain command, otherwise Android might destroy the Surface
                    // too early
                    engine?.flushAndWait()
                    swapChain = null
                }
            }

            override fun onResized(width: Int, height: Int) {
                val aspect = width.toDouble() / height.toDouble()
                camera.setProjection(45.0, aspect, 0.1, 20.0, Camera.Fov.VERTICAL)

                view.viewport = Viewport(0, 0, width, height)

                FilamentHelper.synchronizePendingFrames(engine)
                Log.i("427", uiHelper.isReadyToRender.toString());
                if (uiHelper.isReadyToRender) {
                    // 执行渲染操作
                    swapChain?.let { renderer.beginFrame(it, System.nanoTime()) }
                    renderer.render(view)
                    renderer.endFrame()
                }
            }
        }

        uiHelper.attachTo(surfaceView)
    }

but the UiHelper.RendererCallback never executed, I think it is due to uiHelper.isReadyToRender is always false, and thus the swapChain is always null. Is it because of the uiHelper.isReadyToRender is false that  UiHelper.RendererCallback never executed.

Thanks!

**Comment by yicong123456** [comment]
It shows that uiHelper.isReadyToRender is true when mHasSwapChain is true in source code. Not really sure the relationship between createSwapChain in UiHelper.java and Engin.java. However it seems that mHasSwapChain will be false thus uiHelper.isReadyToRender will also be false then I cannot get back to the onNativeWindowChanged callback then I can not set swapChain by swapChain = engine?.createSwapChain(surface) inside that function.
